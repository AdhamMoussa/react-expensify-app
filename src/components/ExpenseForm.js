import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";

class ExpenseForm extends React.Component {
  state = {
    description: this.props.editedExpense
      ? this.props.editedExpense.description
      : "",
    amount: this.props.editedExpense ? this.props.editedExpense.amount : "",
    note: this.props.editedExpense ? this.props.editedExpense.note : "",
    createdAt: this.props.editedExpense
      ? this.props.editedExpense.createdAt
      : moment(),
    datePickerFocused: false,
    descriptionError: null,
    amountError: null
  };
  handleDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  handleAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
      this.setState(() => ({ amount }));
  };
  handleNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  handleFormSubmit = e => {
    e.preventDefault();
    // handle description error
    if (!this.state.description) {
      this.setState(() => ({
        descriptionError: "Please insert a description of your expense"
      }));
    } else if (this.state.description && this.state.descriptionError) {
      this.setState(() => ({ descriptionError: null }));
    }
    // handle amount error
    if (!this.state.amount) {
      this.setState(() => ({
        amountError: "Please insert an amount of your expense"
      }));
    } else if (this.state.amount && this.state.amountError) {
      this.setState(() => ({ amountError: null }));
    }
    // Submit
    if (this.state.description && this.state.amount) {
      this.setState(() => ({
        descriptionError: null,
        amountError: null
      }));
      this.props.onSubmitHandler({
        description: this.state.description,
        amount: parseFloat(this.state.amount),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <div>
        <form onSubmit={this.handleFormSubmit}>
          <input
            id="description-input"
            type="text"
            name="description"
            value={this.state.description}
            onChange={this.handleDescriptionChange}
            placeholder="Expense Description"
            autoFocus
          />
          <br />
          {this.state.descriptionError && <p>{this.state.descriptionError}</p>}
          <input
            id="amount-input"
            type="text"
            name="amount"
            value={this.state.amount}
            onChange={this.handleAmountChange}
            placeholder="Amount"
          />
          <br />
          {this.state.amountError && <p>{this.state.amountError}</p>}
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={date => {
              if (date) {
                this.setState({ createdAt: date });
              }
            }}
            focused={this.state.datePickerFocused}
            onFocusChange={({ focused }) =>
              this.setState({ datePickerFocused: focused })
            }
            numberOfMonths={1}
            isOutsideRange={day => false}
          />
          <br />
          <textarea
            id="note-input"
            name="note"
            value={this.state.note}
            onChange={this.handleNoteChange}
            placeholder="Expense Note"
          />
          <br />
          <button type="submit">Submit</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
