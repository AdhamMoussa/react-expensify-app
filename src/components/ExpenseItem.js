import React from "react";
import { Link } from 'react-router-dom';
import numeral from 'numeral';

const ExpenseItem = (props) => {
  return (
    <div>
      <Link to={`/edit/${props.id}`}>
        <h3>Expense: {props.description}</h3>
      </Link>
      <p>Amount: {numeral(props.amount).format('$0,0.00')}</p>
      <p>Date: {props.createdAt.format('MMM Do, YYYY')}</p>
    </div>
  );
};

export default ExpenseItem;
