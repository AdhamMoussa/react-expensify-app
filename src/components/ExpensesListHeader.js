import React from "react";
import { connect } from 'react-redux';
import getExpensesView from '../selectors/expenses';
import getTotals from "../selectors/getTotals";
import numeral from "numeral";

export const ExpensesListHeader = props => {
  return (
    <div>
      <p>
        You have {props.totalCount} expense(s), totalling
        {numeral(props.totalAmount).format("$0,0.00")}
      </p>
    </div>
  );
};

const mapStateToProps = (state) => {
  const expensesView = getExpensesView(state.expenses, state.filters);
  return {
    totalAmount: getTotals(expensesView),
    totalCount: expensesView.length
  };
};

export default connect(mapStateToProps)(ExpensesListHeader);
