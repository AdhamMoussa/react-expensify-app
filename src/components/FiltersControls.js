import React from "react";
import { connect } from "react-redux";
import {
  setFilterText,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from "../actions/filters";
import { DateRangePicker } from "react-dates";

export class FiltersControls extends React.Component {
  state = {
    focusedInput: null
  };
  handleFilterTextChange = e => {
    const newValue = e.target.value;
    this.props.setFilterText(newValue);
  };
  handleSortChange = e => {
    if (e.target.value === "date") {
      this.props.sortByDate();
    } else if (e.target.value === "amount") {
      this.props.sortByAmount();
    }
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  render() {
    return (
      <div className="filters-controls">
        <div className="filters-controls__item">
          <input
            id="filter-text-input"
            type="text"
            value={this.props.filters.text}
            onChange={this.handleFilterTextChange}
            placeholder="Search Expenses..."
          />
        </div>
        <div className="filters-controls__item">
          <select
            id="sort-select"
            value={this.props.filters.sortBy}
            onChange={this.handleSortChange}
          >
            <option value="date">Date</option>
            <option value="amount">Amount</option>
          </select>
        </div>
        <div className="filters-controls__item">
          <DateRangePicker
            startDate={this.props.filters.startDate}
            startDateId={"a1b2"}
            endDate={this.props.filters.endDate}
            endDateId={"x1z2"}
            onDatesChange={this.onDatesChange}
            focusedInput={this.state.focusedInput}
            onFocusChange={focusedInput => this.setState({ focusedInput })}
            numberOfMonths={1}
            isOutsideRange={day => false}
            showClearDates={true}
          />
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ filters }) => ({
  filters
});

const mapDispatchToProps = dispatch => {
  return {
    setFilterText: text => dispatch(setFilterText(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: startDate => dispatch(setStartDate(startDate)),
    setEndDate: endtDate => dispatch(setEndDate(endtDate))
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(FiltersControls);
