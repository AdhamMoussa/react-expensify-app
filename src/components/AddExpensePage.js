import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startAddExpense } from '../actions/expenses';


export const AddExpensePage = (props) => {
  const handleFormSubmit = (expense) => {
    props.onSubmit(expense);
    props.history.push('/');
  };
  return (
    <div>
      <h2>Add Expense</h2>
      <ExpenseForm onSubmitHandler={handleFormSubmit} />
    </div>
  );
};

const mapDispatchToProps = (dispatch) => {
  return {
    onSubmit: (expense) => dispatch(startAddExpense(expense))
  };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
