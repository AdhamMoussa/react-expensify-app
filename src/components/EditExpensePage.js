import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { editExpense, removeExpense } from "../actions/expenses";

export const EditExpensePage = props => {
  const onSubmitHandler = expense => {
    props.onSubmit(props.expense.id, expense);
    props.history.push("/");
  };
  const onRemoveHandler = () => {
    props.onRemoveClick(props.expense.id);
    props.history.push("/");
  };
  return (
    <div>
      <h2>Edit Expense</h2>
      <ExpenseForm
        onSubmitHandler={onSubmitHandler}
        editedExpense={props.expense}
      />
      <button id="remove-btn" onClick={onRemoveHandler}>
        Remove
      </button>
    </div>
  );
};

const mapDispatchToProps = dispatch => {
  return {
    onSubmit: (id, expense) => dispatch(editExpense(id, expense)),
    onRemoveClick: id => dispatch(removeExpense(id))
  };
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(exp => exp.id === props.match.params.id)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
