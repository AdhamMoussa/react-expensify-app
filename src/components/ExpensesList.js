import React from "react";
import { connect } from "react-redux";
import getVisibleExpenses from "../selectors/expenses";
import ExpenseItem from "./ExpenseItem";
import FiltersControls from "./FiltersControls";
import ExpensesListHeader from './ExpensesListHeader';

export const ExpensesList = props => {
  return (
    <div>
      <ExpensesListHeader />
      <FiltersControls />
      <div>
        {props.expenses.length === 0 ? (
          <p>No Expenses</p>
        ) : (
          props.expenses.map(exp => {
            return <ExpenseItem {...exp} key={exp.id} />;
          })
        )}
      </div>
    </div>
  );
};

const mapStateToProps = ({ expenses, filters }) => {
  return {
    expenses: getVisibleExpenses(expenses, filters)
  };
};

export default connect(mapStateToProps)(ExpensesList);
