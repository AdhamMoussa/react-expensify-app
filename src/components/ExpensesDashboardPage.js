import React from 'react';
import ExpensesList from './ExpensesList';

const ExpensesDashboardPage = (props) => {
  return (
    <div>
      <h2>Your Expenses</h2>
      <ExpensesList />
    </div>
  )
};


export default ExpensesDashboardPage;
