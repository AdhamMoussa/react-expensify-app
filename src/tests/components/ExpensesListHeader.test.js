import React from "react";
import { shallow } from "enzyme";
import { ExpensesListHeader } from "../../components/ExpensesListHeader";

test("should render ExpensesListHeader correctly without expenses", () => {
  const wrapper = shallow(
    <ExpensesListHeader totalAmount={0} totalCount={0} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpensesListHeader correctly with expenses", () => {
  const wrapper = shallow(
    <ExpensesListHeader totalAmount={1401} totalCount={4} />
  );
  expect(wrapper).toMatchSnapshot();
});
