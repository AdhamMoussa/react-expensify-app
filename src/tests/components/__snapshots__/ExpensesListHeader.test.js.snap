// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesListHeader correctly with expenses 1`] = `
<div>
  <p>
    You have 
    4
     expense(s), totalling
    $1,401.00
  </p>
</div>
`;

exports[`should render ExpensesListHeader correctly without expenses 1`] = `
<div>
  <p>
    You have 
    0
     expense(s), totalling
    $0.00
  </p>
</div>
`;
